utils/src/main/java/software/amazon/awssdk/utils/async/SimplePublisher.java

```diff
@@ -0,0 +1,497 @@
+                        subscriber = null; // Allow subscriber to be garbage collected after cancellation.
+                        break;
+                    default:
+                        // Should never happen. Famous last words?
+                        throw new IllegalStateException("Unknown entry type: " + entry.type());
+                }
+
+                entry.resultFuture.complete(null);
+            }
+
+            eventQueue.remove();
+        }
+    }
+
+    /**
+     * Return true if we should process the provided queue entry.
+     */
+    private boolean shouldProcessQueueEntry(QueueEntry<T> entry) {
+        if (subscriber == null) {
+            // We don't have a subscriber yet.
+            return false;
+        }
+
+        if (entry == null) {
+            // The queue is empty.
+            return false;
+        }
+
+        if (entry.type() != ON_NEXT) {
+            // This event isn't an on-next event, so we don't need subscriber demand in order to process it.
+            return true;
+        }
+
+        if (entryTypesToFail.contains(ON_NEXT)) {
+            // This is an on-next call (decided above), but we're failing on-next calls. Go ahead and fail it.
+            return true;
+        }
+
+        // This is an on-next event and we're not failing on-next events, so make sure we have demand available before
+        // processing it.
+        return outstandingDemand.get() > 0;
+    }
+
+    /**
+     * Invoked from within {@link #processEventQueue()} when we can't process the queue for some reason. This is likely
+     * caused by a downstream subscriber throwing an exception from {@code onNext}, which it should never do.
+     *
+     * <p>Here we try our best to fail all of the entries in the queue, so that no callers have "stuck" futures.
+     */
+    private void panicAndDie(Throwable cause) {
+        try {
+            // Create exception here instead of in supplier to preserve a more-useful stack trace.
+            RuntimeException failure = new IllegalStateException("Encountered fatal error in publisher", cause);
+            rejectException.compareAndSet(null, () -> failure);
+            entryTypesToFail.addAll(asList(QueueEntry.Type.values()));
+            subscriber.onError(failure);
+
+            while (true) {
+                QueueEntry<T> entry = eventQueue.poll();
+                if (entry == null) {
+                    break;
+                }
+                entry.resultFuture.completeExceptionally(failure);
+            }
+        } catch (Throwable t) {
+            t.addSuppressed(cause);
+            log.error(() -> "Failed while processing a failure. This could result in stuck futures.", t);
+        }
+    }
+
+    /**
+     * Ensure that {@link #rejectException} is null. If it is not, throw the exception.
+     */
+    private void checkRejectException() {
+        if (rejectException.get() != null) {
+            throw rejectException.get().get();
```

Comment:
Instead of relying on throw and try-catch to complete the future exceptionally, would it be better to return an `Optional<RuntimeException>` and make the caller handle it more explicitly? Because if future developers forget to catch exceptions when calling this method, it could cause the future to be stuck.

Reply:
Hmm, let me play with it a bit. I'm not sure how common the scenario "developers think to call this, but don't think to catch the exception it throws" is, but it's certainly possible. I was optimizing for readability in the functions that call `checkRejectException`, similar to how `Validate.isTrue` throws an exception instead of returning an `Optional<RuntimeException>`. I wanted to minimize the ceremony in the caller.

Reply:
I couldn't make it into something that I liked. I renamed the method to follow the "validate*" convention so that the caller can hopefully realize that it throws an exception.